README.md
# froxy - Flexible Proxy and vhost

A simple, flexible proxy and vhost library for Node.js.

Easily create HTTP request handlers for proxying to arbitrary hosts.

Features:
* Fast, efficient proxy using request and response pipes.
* Create proxy request handler with a simple config.
* Create vhost dispatch handler with a simple config.
* [Express](http://expressjs.com)-compatible
* Created handler is reusable in multiple servers or Express routes.
* Support setting of origin host, port, protocol
* Support custom connection timeout setting
* Relative URL translation via simple prefix substitution, Regex substitution, or custom JS function
* Loopback proxying when both host and port are omitted


## Installation

```sh
$ npm install froxy
``` 

## Usage

The following snippets are taken from full demo code at [https://github.com/leonardw/froxy-proxy-example](https://github.com/leonardw/froxy-proxy-example)

Simple HTTP proxy
```js
var froxy = require('froxy'),
	http = require("http");

http.createServer(froxy.proxy({
	host: 'en.wikipedia.org',
	debug: true
})).listen(8000);
```

Using Express (v4), implementing a multi-origin proxy depending on incoming URL
```js
var express = require('express'),
    froxy = require('froxy'),
    http = require('http');

var app = express();
app.set('port', 8000);

// Github URLs proxy to github.com/leonardw/
app.route('/github/*').get(froxy.proxy({
    host: 'github.com',
    translate: ['/github/', '/leonardw/']
}));

// All other URLs proxy to Wikipedia
app.route('/*').get(froxy.proxy({
    host: 'en.wikipedia.org'
}));

http.createServer(app).listen(app.get('port'), function(){
    console.log('Express server listening on port', app.get('port'));
});
```


## API

#### .proxy(config)
Creates and returns an HTTP handler that proxies requests to an origin server described by a `config` containing the following properties:
* `host` (optional) : The origin host to proxy to. If omitted, it's the same as the request host.
* `port` (optional) : The origin port to proxy to. When both `host` and `port` are omitted, this acts as a loopback proxy, connecting back on to itself through the same request host and port; if a `host` is provided, this defaults to unset, and hence respect the HTTP implicit defaults, i.e. port 80 for HTTP and 443 for HTTPS.
* `protocol` (optional, `'http'|'https'`) : The protocol to use. If omitted, it's the same as the request protocol.
* `translate` (optional) : A translation rule for the relative path of URL. This may be a two-element array of the form `[prefixPath, tranlatePrefixPath]`
or `[regexPattern, translateTemplate]`, or a custom function with signature `function(url, urlSpec, request)` that returns a translated string URL.
Defaults to no translation, i.e., same as request URL.
* `timeout` (optional) : A timeout, in milliseconds, to wait for response from origin host. Default 5 minutes.
* `debug` (optional, `true|false`) : Set to `true` to log useful debug information to console. Defaults to `false`.


Where `translate` is provided as a `function(url, urlSpec, request)`, the calling parameters are
* `url` : The incoming relative request URL string.
* `urlSpec` : The request URL specification of the form `{ host:, port:, protocol:, secure:, url: }`.
* `request` : The request object as provided by Node.

#### .vhost(config)
#### .vhost().map(config)
#### .vhost().dispatcher
*vhost documentation coming soon...*

##License

(The MIT License)

Copyright (c) 2014 Leonard Wu <leonard.wu92@alumni.ic.ac.uk>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
# Change Log - froxy

### 0.5.2, 2014-09-09
* Fix infinite redirect loop when auto redirecting browser to HTTPS but an intermediate proxy has modified protocol.
* Handle x-forwarded-... headers introduced by intermediate proxy

### 0.5.1, 2014-09-05
* New proxy option 'access' to limit incoming protocol access to plain, secure or any
* vhost map() renamed to reset() to avoid confusion with ECMAScript-5 map function

### 0.5.0, 2014-09-01
* Redesigned vhost API, allowing vhost mapping to be changed on-the-fly.

### 0.4.1, 2014-08-26
* Much improved stability
* Graceful error handling of problematic origin server
* Support timeout on proxy connections
* Consistent HTTP error responses

### 0.2.0, 2014-08-26
* New vhost functionality via .vhost()

### 0.1.0, 2014-07-16
* Fix variable declaration error in JS caused by typo
* Update usage example code to Express v4
* Promoted version number to reflect readiness after initial trials

### 0.0.2, 2014-07-09
First documented release

lib/froxy.js
/*!
 * Froxy
 * Copyright (c) 2014 Leonard Wu <leonard.wu92@alumni.ic.ac.uk>
 * https://github.com/leonardw/froxy
 * MIT Licensed
 */
(function () {
	var request = require('request'),
		http = require('http'),
		util = require('util');
	
	var PKG = require('../package.json').name,
		VER = require('../package.json').version;
	
	var HTTP = {
		FORBIDDEN: 403,
		NOT_FOUND: 404,
		INTERNAL_SERVER_ERROR: 500,
		BAD_GATEWAY: 502,
		SERVICE_UNAVAILABLE: 503,
		GATEWAY_TIMEOUT: 504
	};
	
	var DEFAULT_HTTP_TIMEOUT = 1000 * 60 * 5, //in millisec; 5 min, reference Apache mod_proxy
		DEFAULT_ACCESS = 'any';
	
	function getTranslateFn(translateParam) {
		if (translateParam instanceof Function) {
			// param is already a function, return as is
			return translateParam;
		} else if (util.isArray(translateParam)) {
			// param is an array of either [prefixPath, tranlatePrefixPath] or [regexPattern, translateTemplate]
			if (translateParam[0] instanceof Object) {
				// First param is a regex pattern object, of form [regexPattern, translateTemplate]
				return function(url) {
					try {
						return url.replace(translateParam[0], translateParam[1]);
					} catch (ex) {
						console.error('URL regex translation error:', ex);
					}
					return null;
				};
			} else {
				// First param should be a string, of form [prefixPath, tranlatePrefixPath]
				return function(url) {
					if (url.indexOf(translateParam[0]) === 0) {
						return translateParam[1] + url.substring(translateParam[0].length);
					} else {
						console.error('URL translation failed to match "%s"', translateParam[0]);
						return null;
					}
				};
			}
		} else {
			// no translation will be done
			return null;
		}
	}

	// Get the request specification, i.e. host, port, and protocol
	function getRequestSpec(req) {
		var host_port = req.headers.host.split(':'),
			secure = req.connection.encrypted ? true : false,
			protocol = secure ? 'https' : 'http';
		return {
			host: host_port[0],
			port: host_port[1] ? parseInt(host_port[1], 10) : null,
			protocol: protocol,
			secure: secure,
			url: req.url
		};
	}
	
	// Get the request specification as indicated by 'forward' headers, i.e. host, port, and protocol
	function getSourceRequestSpec(req) {
		var host = req.headers['x-forwarded-for'],
			port = req.headers['x-forwarded-port'],
			protocol = req.headers['x-forwarded-proto'],
			secure = protocol ? (protocol==='https') : null;
		return {
			host: host || null,
			port: port ? parseInt(port, 10) : null,
			protocol: protocol || null,
			secure: secure
		};
	}
	
	function sendResponse(res, status, msgOrCallbk) {
		if (res) {
			var description = http.STATUS_CODES[status];
			if (description === undefined) {
				// unrecognised HTTP code, fallback to defaults
				description = 'Unrecognised HTTP status';
			}
			res.writeHead(status);
			res.write('<html><head><title>' + PKG + ': ' + description + '</title></head><body>');
			res.write('<div style="background-color:#3399FF; color:#FFFFFF; text-align:right; padding:5px 10px;">');
			res.write(PKG + ' v' + VER);
			res.write('</div>');
			res.write('<h1>' + status + ' - ' + description + '</h1>');
			if (msgOrCallbk) {
				if (msgOrCallbk instanceof Function) {
					// we have a callback, call it now with response obj
					msgOrCallbk(res);
				} else {
					// it's a message
					res.write('<p>' + msgOrCallbk + '</p>');
				}
			}
			res.write('</body></html>');
			res.end();
		} else {
			console.error('Unable to send HTTP response - invalid response object');
		}
	}
	
	function constructUrl(proto, host, port, relUrl) {
		var vPort = port;
		if ( proto==='http' && port === 80 ) {
			vPort = null;
		} else if ( proto==='https' && port === 443 ) {
			vPort = null;
		}
		var fqUrl = proto + "://" + host + (vPort? ':'+vPort : '') + (relUrl? relUrl : '');
		return fqUrl;
	}

	// Given config options, creates a handler for proxying requests
	function proxy(option) {
		// Set defaults
		var cfg = {
			host: option.host || null,
			port: option.port || null,
			protocol: option.protocol || null,
			translate: getTranslateFn(option.translate),
			timeout: option.timeout || DEFAULT_HTTP_TIMEOUT,
			access: (option.access === 'plain' || option.access === 'secure' || option.access === 'any')? option.access : DEFAULT_ACCESS,
			debug: option.debug === true
		};
		
		// Return the request handler function
		return function(req, res) {
			var reqSpec = getRequestSpec(req);
			if (cfg.debug) {
				console.info('Request:', reqSpec);
			}
			if (cfg.access) {
				if (cfg.access === 'secure') {
					var fwdReqSpec = getSourceRequestSpec(req);
					if ( !((fwdReqSpec.secure === null) ? reqSpec.secure : fwdReqSpec.secure) ) {
						// only allow secure access; redirect to HTTPS
						var redirectLoc = constructUrl('https', reqSpec.host, null, reqSpec.url);
						res.writeHead(302, {Location: redirectLoc});
						res.end();
						return;
					}
				} else if (reqSpec.secure) {
					// only allow plain access; return 403 Forbidden
					sendResponse(res, HTTP.FORBIDDEN);
					return;
				}
			}
			// Translate the relative part of URL
			var relativeUrl = (cfg.translate? cfg.translate(reqSpec.url, reqSpec, req) : reqSpec.url);
			if (relativeUrl === null) {
				console.error('Failed to translate URL:', reqSpec.url);
				// Failed to translate, return 404 Not Found
				sendResponse(res, HTTP.NOT_FOUND);
			} else {
				// Build Fully-qualified URL for rewrite
				var fqUrl = (cfg.protocol ? cfg.protocol : reqSpec.protocol) + "://"
					+ (cfg.host ? cfg.host : reqSpec.host)
					+ (cfg.port ? ':'+cfg.port : (cfg.host ? '' : (reqSpec.port ? ':'+reqSpec.port : '')))
					+ relativeUrl;
				if (cfg.debug) {
					console.info('URL    :', reqSpec.url);
					console.info('Origin :', fqUrl);
					console.info('');
				}
				// Proxy request to the translated URL, and pipe back the response
				req.pipe(request(
					{
						url:fqUrl,
						timeout:cfg.timeout
					},
					function(rErr, httpMsg, rBody){
						if (rErr) {
							if (rErr.code === 'ECONNREFUSED'){
								console.error('%s: Connection refused by host', PKG, fqUrl);
								sendResponse(res, HTTP.SERVICE_UNAVAILABLE);
							} else if (rErr.code === 'ECONNRESET'){
								console.error('%s: Connection reset by host', PKG, fqUrl);
								sendResponse(res, HTTP.BAD_GATEWAY);
							} else if (rErr.code === 'ETIMEDOUT'){
								console.error('%s: Connection timeout while trying host', PKG, fqUrl);
								sendResponse(res, HTTP.GATEWAY_TIMEOUT);
							} else {
								console.error('%s: Internal server error %s: %s while connecting to', PKG, rErr.code, rErr, fqUrl);
								sendResponse(res, HTTP.INTERNAL_SERVER_ERROR);
							}
						}
					}
				)).pipe(res);
			}
			return;
		};
	}
	
	// Default internal vhost handler
	// Lists all recognised vhosts
	function vhostListingHandler(vhosts) {
		return function(req, res) {
			var reqSpec = getRequestSpec(req);
			var fwdReqSpec = getSourceRequestSpec(req);
			sendResponse(res, HTTP.NOT_FOUND, function(res){
				res.write('<p>Unrecognised vhost: ' + reqSpec.host + '</p>');
				res.write('<p>');
				var firstEntry = true;
				vhosts.forEach(function(hostname) {
					if (firstEntry) {
						res.write('Did you mean:<br/>');
						firstEntry = false;
					}
					var url = constructUrl(
						fwdReqSpec.protocol || reqSpec.protocol,
						hostname,
						fwdReqSpec.port || reqSpec.port,
						reqSpec.url
					);
					res.write('<a href="' + url + '">' + url +'</a><br/>');
				});
				res.write('</p>');
			});
		};
	}
	
	function vhost(option) {
		var vhostMap = {},
			defaultHandler;
		
		// vhost dispatcher function
		function dispatcher(req, res) {
			var hostname = req.headers.host.split(':')[0];
			var handler = vhostMap[hostname];
			if (handler) {
				handler(req, res);
			} else {
				defaultHandler(req, res);
			}
		}
		
		function reset(option) {
			defaultHandler = null;
			Object.keys(option).forEach(function(hostname) {
				var val = option[hostname],
					handler = null;
				if (val instanceof Function) {
					// handler function provided, use as is
					handler = val;
				} else if (val instanceof Object && !util.isArray(val)) {
					// create handler from proxy config
					handler = proxy(val);
				}
				if (handler) {
					if (hostname === 'default') {
						defaultHandler = handler;
					} else {
						vhostMap[hostname] = handler;
					}
				}
			});
			if (!defaultHandler) {
				// No default vhost handler provided
				// Use internal default
				defaultHandler = vhostListingHandler(Object.keys(vhostMap));
			}
		}
		
		reset(option || {});
		
		return {
			reset: reset,
			dispatcher: dispatcher
		};
	}
	
	var _api = {
		proxy: proxy,
		vhost: vhost
	};
	
	if (typeof module !== 'undefined' && module.exports) {
		// Running in Node.js
		module.exports = _api;
	} else {
		// We are lost
		console.error('Unknown execution environment. Giving up.');
	}
	
})();
